Introduce allow_redirects flag to control redirect behaviour of requests

cynthia2006 opened the pull request:
<!--
    **IMPORTANT**: PRs without the template will be CLOSED
    
    Due to the high volume of pull requests, it may be a while before your PR is reviewed.
    Please try to keep your pull request focused on a single bugfix or new feature.
    Pull requests with a vast scope and/or very large diff will take much longer to review.
    It is recommended for new contributors to stick to smaller pull requests, so you can receive much more immediate feedback as you familiarize yourself with the codebase.

    PLEASE AVOID FORCE-PUSHING after opening a PR, as it makes reviewing more difficult.
-->

### Description of your *pull request* and other information


Fixes #14542 by including a boolean flag `allow_redirects` which defaults to `True` for `_download_webpage()`, `_download_webpage_handle()`, `_request_webpage()` methods, and `yt_dlp.networking.Requests` constructor, etc.


<details open><summary>Template</summary> <!-- OPEN is intentional -->

<!--
    # PLEASE FOLLOW THE GUIDE BELOW

    - You will be asked some questions, please read them **carefully** and answer honestly
    - Put an `x` into all the boxes `[ ]` relevant to your *pull request* (like [x])
    - Use *Preview* tab to see what your *pull request* will actually look like
-->

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check those that apply and remove the others:
- [x] I am the original author of the code in this PR, and I am willing to release it under [Unlicense](http://unlicense.org/)
- [x] I am not the original author of the code in this PR, but it is in the public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)
- [x] I have read the [policy against AI/LLM contributions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#automated-contributions-ai--llm-policy) and understand I may be blocked from the repository if it is violated

### What is the purpose of your *pull request*? Check those that apply and remove the others:
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [x] Core bug fix/improvement
- [x] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

cynthia2006 committed:
Introduce allow_redirects flag to control redirects on a per-request basis

coletdjnz commented on yt_dlp/extractor/common.py:
```diff
@@ -847,7 +847,7 @@ def __can_accept_status_code(err, expected_status):
             return err.status in variadic(expected_status)
 
-    def _create_request(self, url_or_request, data=None, headers=None, query=None, extensions=None):
+    def _create_request(self, url_or_request, data=None, headers=None, query=None, extensions=None, allow_redirects=True):
```
We want to avoid adding parameters that are rarely used to these functions, so I would instead in the caller create a Request object with the allow_redirects extension set.

coletdjnz reviewed (changes requested):
pls review the PR checkboxes

coletdjnz commented:
Also add test(s) in test_networking.py (there should be some redirect tests you can derive from)

p.s avoid force pushing as it messes with reviews 😅

cynthia2006 commented:
I'm sorry, but I had some other changes accidentally pushed.

cynthia2006 committed:
Turn allow_redirects into a Request extension

This also considers the case where Requests consumes the response
body, leaving nothing in the Urllib3 raw response -- problematic
when allow_redirects=False and the HTTP status code is 3xx.

dirkf commented:
What is the expected behaviour with `_download_webpage(..., fatal=True)` (etc) when a 3xx code is received in response to a request that specifies not allowing redirects?

Should the call raise `HTTPError`, so that getting the non-redirect page means specifying, eg, `expected_status=(301, 302, 303, 307, 308)`, or calling `_download_webpage_handle()` and catching the exception? Or should the plain method succeed with the redirect code only visible by calling `_download_webpage_handle()` (probably more useful)?

cynthia2006 commented:
> What is the expected behaviour with `_download_webpage(..., fatal=True)` (etc) when a 3xx code is received in response to a request that specifies not allowing redirects?
> 
> Should the call raise `HTTPError`, so that getting the non-redirect page means specifying, eg, `expected_status=(301, 302, 303, 307, 308)`, or calling `_download_webpage_handle()` and catching the exception? Or should the plain method succeed with the redirect code only visible by calling `_download_webpage_handle()` (probably more useful)?

Since, `allow_redirects` is a `Requests` extension, it'd be usable from anything that accepts a object instance of that base class. The expected behaviour of `YoutubeDL.urlopen()` to raise a `HTTPError`, so that `download_webpage_handle()` thereby `download_webpage()` would catch the errors if the expected 3xx status codes were specified explicitly. 

Although, as a simplification `download_webpage()` may have these as the default if the Requests object had the `allow_redirects` extension in it — what do you think? @dirkf

dirkf commented:
> ... what do you think? 

The `HTTPRedirectHandler` class in `urllib.request` and Py2 `urllib2` has a `max_redirections` attribute (default value 10) that causes `HTTPError` when exceeded. `requests.Session` has a similar `max_redirects` attribute (default 30). `urllib3.request()` has a fancy multi-valued `retries` parameter (default 3) that also controls the number of redirections when they are enabled, AIUI.  And from the PR code, the default is 5 for `curlcffi`.[^1] 

Since this is an edge usage, maybe we should just accept that 3xx errors could be raised. A named constant for the tuple of known/supported 3xx codes would make it easy to use `expected_status` to force returning the unredirected page. It would be good for `HTTPError` to be raised regardless of what any underlying library does. 

Anyhow it's up to you and _yt-dlp_ members, as long as it's consistent.

[^1]: These defaults could be harmonised in yt_dlp.networking. If they're not, maybe we don't care.

dirkf commented on yt_dlp/networking/_urllib.py:
```diff
@@ -388,6 +388,7 @@ def _send(self, request):
             data=request.data,
             headers=headers,
             method=request.method,
+
```
So for urllib2/urllib, it should be possible to specialise the error handler, by replacing l.216 of the `RedirectHandler` with this (de-compat version for Py3 urllib.request):
```py
    def http_error_302(self, req, fp, code, msg, headers):
        allow_redirects = req.extensions.get('allow_redirects') is not False
        if not allow_redirects:
            # could have .request.addinfourl, but .response is where it's defined 
            resp = urllib.response.addinfourl(fp, headers, req.get_full_url(), code)
            resp.msg = msg
            return resp
        # actually could have super().http_error_302(...) since the super methods are also all identical, but that's internal to urllib.request
        return getattr(super(), 'http_error_%s' % (code,))(
            req, fp, code, msg, headers)

    http_error_301 = http_error_303 = http_error_307 = http_error_308 = http_error_302
```
